Correctly calculate doc count error at the slice level for concurrent segment search

jed326 opened the pull request:
### Description
This PR fixes how doc count error is calculated for terms aggregations for conurrent segment search and sets the `slice_size` heuristic introduced in #11585 to be equal to the `shard_size`.

The crux of this change is the introduction of `hasSliceLevelDocCountError`, to indicate to the coordinator whether or not the `slice_size` is the reason for why the error exists. This affects how the top level aggregation error is then calculated in the single shard (and single slice) scenarios.

### Related Issues
Resolves #11680
Resolves #11702

### Check List
- [x] New functionality includes testing.
  - [x] All tests pass
- [x] New functionality has been documented.
  - [x] New functionality has javadoc added
- [x] Failing checks are inspected and point to the corresponding known issue(s) (See: [Troubleshooting Failing Builds](../blob/main/CONTRIBUTING.md#troubleshooting-failing-builds))
- [x] Commits are signed per the DCO using --signoff
- [x] Commit changes are listed out in CHANGELOG.md file (See: [Changelog](../blob/main/CONTRIBUTING.md#changelog))
- [x] Public documentation issue/PR [created](https://github.com/opensearch-project/documentation-website/issues/new/choose)

By submitting this pull request, I confirm that my contribution is made under the terms of the Apache 2.0 license.
For more information on following Developer Certificate of Origin and signing off your commits, please check [here](https://github.com/opensearch-project/OpenSearch/blob/main/CONTRIBUTING.md#developer-certificate-of-origin).

jed326 commented:
@sohami @reta could you please help review? Thanks!

sohami commented on server/src/internalClusterTest/java/org/opensearch/search/aggregations/bucket/TermsShardMinDocCountIT.java:
```diff
@@ -157,6 +161,10 @@ private void addTermsDocs(String term, int numInClass, int numNotInClass, List<I
     // see https://github.com/elastic/elasticsearch/issues/5998
     public void testShardMinDocCountTermsTest() throws Exception {
+        assumeFalse(
+            "For concurrent segment search there will be additional error introduced during the slice level reduce and thus different buckets may be returned See https://github.com/opensearch-project/OpenSearch/issues/11680",
```
same as above

sohami commented on server/src/internalClusterTest/java/org/opensearch/search/aggregations/bucket/TermsShardMinDocCountIT.java:
```diff
@@ -88,6 +88,10 @@ private static String randomExecutionHint() {
 
     // see https://github.com/elastic/elasticsearch/issues/5998
     public void testShardMinDocCountSignificantTermsTest() throws Exception {
+        assumeFalse(
```
this test is validating the count of buckets returned with size and shardSize parameter. I would think that bucket count should be same in both the cases ?

jed326 replied:
Significant terms aggregations return the top `size` buckets by significance score. The `min_doc_count` and it's shard counterpart are therefore used to calibrate the "certainty" that a significant term is in fact significant. For example, using a higher `shard_min_doc_count` could help filter out typos in the term field.

So for this significant terms test if we increase the `shard_size` then the shards will return more buckets and some buckets may have doc count greater than `min_doc_count`, so they will get returned by the coordinator and fail the first set of assertions.

sohami commented on server/src/internalClusterTest/java/org/opensearch/search/aggregations/bucket/TermsDocCountErrorIT.java:
```diff
@@ -1106,6 +1106,11 @@ public void testDoubleValueFieldSubAggDesc() throws Exception {
+        // Force merge each shard down to 1 segment to verify results are the same between concurrent and non-concurrent search paths
+        // Else for concurrent segment search there will be additional error introduced during the slice level reduce and thus different
+        // buckets may be returned. See https://github.com/opensearch-project/OpenSearch/issues/11680",
+        client().admin().indices().prepareForceMerge("idx_fixed_docs_0", "idx_fixed_docs_1", "idx_fixed_docs_2").get();
```
With force merging segments to 1, concurrent search path will have single slice.  Can we update the `shard_size` parameter instead to see if the returned results are same ?

jed326 replied:
Without any additional modifications to this test we won't be able to make it work with just increasing the shard size. The assertions below ([for example](https://github.com/opensearch-project/OpenSearch/pull/11732/files/1c4c7ef68904120c7d4b1745ffc7d7808a46b235#diff-954b334a87adb1b7244d5ac449c329e9b5a274c8147c890f3dba3f8709400ad9R1126-R1131)), are checking for a specific doc count error which will change when we modify the shard size.

If we make the shard size > cardinality so we collect all the documents we can get rid of the doc count error and have the assertions work for both concurrent and non-concurrent path but that seems to go against the spirit of this test since it is testing that the doc count error is calculated correctly.

jed326 replied:
I wrote up a more detailed explanation for the different possible cases here: https://github.com/opensearch-project/OpenSearch/pull/11732#issuecomment-1883786431.

jed326 commented:
With the changes in this PR, there are basically 3 different cases when comparing the results from non-concurrent search vs. concurrent search. I'll illustrate this with the `ShardSizeTermsIT::testShardSizeEqualsSizeString` test. First, the baseline is non-concurrent search, where the response will always look like so:
```
{
    "keys": {
        "doc_count_error_upper_bound": 5,
        "sum_other_doc_count": 7,
        "buckets": [{
            "key": "1",
            "doc_count": 8,
            "doc_count_error_upper_bound": 0
        }, {
            "key": "3",
            "doc_count": 8,
            "doc_count_error_upper_bound": 0
        }, {
            "key": "2",
            "doc_count": 4,
            "doc_count_error_upper_bound": 2
        }]
    }
}
```

Now, there are basically 3 ways that using concurrent search with the `slice_size` heuristic can change the results.

## 1. Some buckets may have a lower `doc_count` with higher `doc_count_error`. 
This is the most intuitive outcome -- because we apply a `slice_size` we eliminate more buckets at the slice level which will contribute to a greater doc count error.
```
{
    "keys": {
        "doc_count_error_upper_bound": 5,
        "sum_other_doc_count": 8,
        "buckets": [{
            "key": "1",
            "doc_count": 8,
            "doc_count_error_upper_bound": 4
        }, {
            "key": "3",
            "doc_count": 8,
            "doc_count_error_upper_bound": 1
        }, {
            "key": "2",
            "doc_count": 3,
            "doc_count_error_upper_bound": 4
        }]
    }
}
```

## 2. Buckets may have the same `doc_count` but with higher `doc_count_error`.
This case is confusing at first because if we collect all of the existing documents for a bucket key, we would initially expect the `doc_count_error` to be 0. However, that's not the case -- take this example:
```
{
    "keys": {
        "doc_count_error_upper_bound": 6,
        "sum_other_doc_count": 7,
        "buckets": [{
            "key": "1",
            "doc_count": 8,
            "doc_count_error_upper_bound": 5
        }, {
            "key": "3",
            "doc_count": 8,
            "doc_count_error_upper_bound": 5
        }, {
            "key": "2",
            "doc_count": 4,
            "doc_count_error_upper_bound": 3
        }]
    }
}
```

If bucket key "1" has 8 docs across 2 shards, and but in each shard all of the key "1" docs are in the same slice, then we will need to calculate the `doc_count_error` from the slices where bucket key "1" is not in the top `slice_size` buckets.

## 3. Buckets may have a greater `doc_count`
For example:
```
{
    "keys": {
        "doc_count_error_upper_bound": 4,
        "sum_other_doc_count": 6,
        "buckets": [{
            "key": "1",
            "doc_count": 8,
            "doc_count_error_upper_bound": 1
        }, {
            "key": "3",
            "doc_count": 8,
            "doc_count_error_upper_bound": 3
        }, {
            "key": "2",
            "doc_count": 5,
            "doc_count_error_upper_bound": 3
        }]
    }
}
```

This case fundamentally has to do with how "ties" are broken when terms have the same doc count. The tie breaker (in cases where doc count error is relevant), is the bucket key itself. So take a case where bucket key "2" has 1 document on a shard, and bucket key "4" has 2 documents on a shard.
```
shard [4, 4, 2]
slice_1 [4, 2]
slice_2 [4]
```
Now, let's imagine a term agg with `size` == `shard_size` == 1. In non-concurrent search bucket "4" will always be returned ahead of bucket "2", however in concurrent search if the 2 documents for bucket "4" are split across different slices, then bucket "2" will actually be return ahead of bucket "4"!

sohami commented on server/src/internalClusterTest/java/org/opensearch/search/aggregations/bucket/ShardSizeTermsIT.java:
```diff
@@ -355,20 +361,23 @@ public void testShardSizeEqualsSizeDouble() throws Exception {
         for (Terms.Bucket bucket : buckets) {
-            assertThat(bucket.getDocCount(), equalTo(expected.get(bucket.getKeyAsNumber().intValue())));
+            expectedDocCount = expected.get(bucket.getKeyAsNumber().intValue());
+            assertTrue((bucket.getDocCount() == expectedDocCount) || bucket.getDocCount() + bucket.getDocCountError() >= expectedDocCount);
```
lets add a reference to your explanation here which can help to understand why docCount can vary and this assertion makes sense here for concurrent flow.

jed326 commented:
@sohami @reta I updated the PR and also added some write ups related to the big issues we're seeing here. Both are referenced by code comments as well. See:
- https://github.com/opensearch-project/OpenSearch/issues/11847
- https://github.com/opensearch-project/OpenSearch/issues/11680

jed326 commented:
Created a new docs issue to add a lot of the details discussed here to the opensearch docs. The terms and sig terms docs are very lacking currently.

reta commented on server/src/internalClusterTest/java/org/opensearch/search/aggregations/bucket/ShardSizeTermsIT.java:
```diff
@@ -86,20 +86,24 @@ public void testShardSizeEqualsSizeString() throws Exception {
         Terms terms = response.getAggregations().get("keys");
         List<? extends Terms.Bucket> buckets = terms.getBuckets();
-        assertThat(buckets.size(), equalTo(3));
+        assertEquals(3, buckets.size());
```
@jed326 why we are changing these asserts? `assertThat` prints significantly better error message

jed326 replied:
This `assertThat` is `org.junit.Assert.assertThat` which is deprecated, so I started changing this in the methods I touched. We could switch to `org.hamcrest.MatcherAssert.assertThat`, which is what junit recommends, or I can just revert this change.

It's probably easier to make that hamcrest change in all places at once in a big batch so I can just revert this instead.

reta replied:
> It's probably easier to make that hamcrest change in all places at once in a big batch so I can just revert this instead.

Please, thank you

jed326 replied:
Already done!

jed326 commented:
> ❌ Gradle check result for [7571351](https://github.com/opensearch-project/OpenSearch/commit/757135112f508d2e22feb8b01634c75b8a9bc2d0): [FAILURE](https://build.ci.opensearch.org/job/gradle-check/32018/)
> 
> Please examine the workflow log, locate, and copy-paste the failure(s) below, then iterate to green. Is the failure [a flaky test](https://github.com/opensearch-project/OpenSearch/blob/main/DEVELOPER_GUIDE.md#flaky-tests) unrelated to your change?

Looks like there are a few more tests I need to fix actually

jed326 commented:
> https://build.ci.opensearch.org/job/gradle-check/32022/
- https://github.com/opensearch-project/OpenSearch/issues/10704
